File: airbyte-server/src/main/java/io/airbyte/server/handlers/api_domain_mapping/StreamStatusesMapper.java
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.server.handlers.api_domain_mapping;

import static java.time.ZoneOffset.UTC;

import com.fasterxml.jackson.databind.JsonNode;
import io.airbyte.api.model.generated.Pagination;
import io.airbyte.api.model.generated.StreamStatusCreateRequestBody;
import io.airbyte.api.model.generated.StreamStatusIncompleteRunCause;
import io.airbyte.api.model.generated.StreamStatusJobType;
import io.airbyte.api.model.generated.StreamStatusListRequestBody;
import io.airbyte.api.model.generated.StreamStatusRateLimitedMetadata;
import io.airbyte.api.model.generated.StreamStatusRead;
import io.airbyte.api.model.generated.StreamStatusRunState;
import io.airbyte.api.model.generated.StreamStatusUpdateRequestBody;
import io.airbyte.commons.json.Jsons;
import io.airbyte.db.instance.jobs.jooq.generated.enums.JobStreamStatusIncompleteRunCause;
import io.airbyte.db.instance.jobs.jooq.generated.enums.JobStreamStatusJobType;
import io.airbyte.db.instance.jobs.jooq.generated.enums.JobStreamStatusRunState;
import io.airbyte.server.repositories.StreamStatusesRepository;
import io.airbyte.server.repositories.domain.StreamStatus;
import io.airbyte.server.repositories.domain.StreamStatusRateLimitedMetadataRepositoryStructure;
import jakarta.inject.Singleton;
import java.time.Instant;
import java.time.OffsetDateTime;

/**
 * Maps between the API and Persistence layers. It is not static to be injectable and enable easier
 * testing in dependents.
 */
@Singleton
public class StreamStatusesMapper {

  // API to Domain
  public StreamStatus map(final StreamStatusCreateRequestBody api) {
    final var domain = new StreamStatus.StreamStatusBuilder()
        .runState(map(api.getRunState()))
        .transitionedAt(fromMills(api.getTransitionedAt()))
        .workspaceId(api.getWorkspaceId())
        .connectionId(api.getConnectionId())
        .jobId(api.getJobId())
        .jobType(map(api.getJobType()))
        .attemptNumber(api.getAttemptNumber())
        .streamNamespace(api.getStreamNamespace())
        .streamName(api.getStreamName());

    if (null != api.getIncompleteRunCause()) {
      domain.incompleteRunCause(map(api.getIncompleteRunCause()));
    }

    if (null != api.getMetadata()) {
      domain.metadata(map(api.getMetadata()));
    }

    return domain.build();
  }

  public StreamStatus map(final StreamStatusUpdateRequestBody api) {
    final var domain = new StreamStatus.StreamStatusBuilder()
        .runState(map(api.getRunState()))
        .transitionedAt(fromMills(api.getTransitionedAt()))
        .workspaceId(api.getWorkspaceId())
        .connectionId(api.getConnectionId())
        .jobId(api.getJobId())
        .jobType(map(api.getJobType()))
        .attemptNumber(api.getAttemptNumber())
        .streamNamespace(api.getStreamNamespace())
        .streamName(api.getStreamName())
        .id(api.getId());

    if (null != api.getIncompleteRunCause()) {
      domain.incompleteRunCause(map(api.getIncompleteRunCause()));
    }

    if (null != api.getMetadata()) {
      domain.metadata(map(api.getMetadata()));
    }

    return domain.build();
  }

  public JobStreamStatusJobType map(final StreamStatusJobType apiEnum) {
    return apiEnum != null ? JobStreamStatusJobType.lookupLiteral(apiEnum.name().toLowerCase()) : null;
  }

  public JobStreamStatusRunState map(final StreamStatusRunState apiEnum) {
    return apiEnum != null ? JobStreamStatusRunState.lookupLiteral(apiEnum.name().toLowerCase()) : null;
  }

  public JobStreamStatusIncompleteRunCause map(final StreamStatusIncompleteRunCause apiEnum) {
    return apiEnum != null ? JobStreamStatusIncompleteRunCause.lookupLiteral(apiEnum.name().toLowerCase()) : null;
  }

  public StreamStatusesRepository.Pagination map(final Pagination api) {
    return api != null ? new StreamStatusesRepository.Pagination(
        api.getRowOffset() / api.getPageSize(),
        api.getPageSize()) : null;
  }

  public StreamStatusesRepository.FilterParams map(final StreamStatusListRequestBody api) {
    return new StreamStatusesRepository.FilterParams(
        api.getWorkspaceId(),
        api.getConnectionId(),
        api.getJobId(),
        api.getStreamNamespace(),
        api.getStreamName(),
        api.getAttemptNumber(),
        map(api.getJobType()),
        map(api.getPagination()));
  }

  // Domain to API
  public StreamStatusRead map(final StreamStatus domain) {
    final var api = new StreamStatusRead();
    api.setId(domain.getId());
    api.setConnectionId(domain.getConnectionId());
    api.setWorkspaceId(domain.getWorkspaceId());
    api.setJobId(domain.getJobId());
    api.setJobType(map(domain.getJobType()));
    api.setAttemptNumber(domain.getAttemptNumber());
    api.setStreamName(domain.getStreamName());
    api.setStreamNamespace(domain.getStreamNamespace());
    api.setTransitionedAt(domain.getTransitionedAt().toInstant().toEpochMilli());
    api.setRunState(map(domain.getRunState()));

    if (null != domain.getIncompleteRunCause()) {
      api.setIncompleteRunCause(map(domain.getIncompleteRunCause()));
    }

    if (null != domain.getMetadata()) {
      api.setMetadata(map(domain.getMetadata()));
    }

    return api;
  }

  public StreamStatusJobType map(final JobStreamStatusJobType domainEnum) {
    return domainEnum != null ? StreamStatusJobType.fromValue(domainEnum.name().toUpperCase()) : null;
  }

  public StreamStatusRunState map(final JobStreamStatusRunState domainEnum) {
    return domainEnum != null ? StreamStatusRunState.fromValue(domainEnum.name().toUpperCase()) : null;
  }

  public StreamStatusIncompleteRunCause map(final JobStreamStatusIncompleteRunCause domainEnum) {
    return domainEnum != null ? StreamStatusIncompleteRunCause.fromValue(domainEnum.name().toUpperCase()) : null;
  }

  public StreamStatusRateLimitedMetadata map(final JsonNode rateLimitedMetadata) {
    final StreamStatusRateLimitedMetadataRepositoryStructure rateLimitedInfo =
        Jsons.object(rateLimitedMetadata, StreamStatusRateLimitedMetadataRepositoryStructure.class);
    return new StreamStatusRateLimitedMetadata().quotaReset(rateLimitedInfo.getQuotaReset());
  }

  public JsonNode map(final StreamStatusRateLimitedMetadata rateLimitedMetadata) {
    final StreamStatusRateLimitedMetadataRepositoryStructure streamStatusRateLimitedMetadataRepositoryStructure =
        new StreamStatusRateLimitedMetadataRepositoryStructure(rateLimitedMetadata.getQuotaReset());
    return Jsons.jsonNode(streamStatusRateLimitedMetadataRepositoryStructure);
  }

  OffsetDateTime fromMills(final Long millis) {
    return OffsetDateTime.ofInstant(Instant.ofEpochMilli(millis), UTC);
  }

}

File: airbyte-server/src/main/kotlin/io/airbyte/server/handlers/apidomainmapping/StreamStatusesMapper.kt
/*
 * Copyright (c) 2020-2025 Airbyte, Inc., all rights reserved.
 */

package io.airbyte.server.handlers.apidomainmapping

import com.fasterxml.jackson.databind.JsonNode
import io.airbyte.api.model.generated.Pagination
import io.airbyte.api.model.generated.StreamStatusCreateRequestBody
import io.airbyte.api.model.generated.StreamStatusIncompleteRunCause
import io.airbyte.api.model.generated.StreamStatusJobType
import io.airbyte.api.model.generated.StreamStatusListRequestBody
import io.airbyte.api.model.generated.StreamStatusRateLimitedMetadata
import io.airbyte.api.model.generated.StreamStatusRead
import io.airbyte.api.model.generated.StreamStatusRunState
import io.airbyte.api.model.generated.StreamStatusUpdateRequestBody
import io.airbyte.commons.json.Jsons
import io.airbyte.db.instance.jobs.jooq.generated.enums.JobStreamStatusIncompleteRunCause
import io.airbyte.db.instance.jobs.jooq.generated.enums.JobStreamStatusJobType
import io.airbyte.db.instance.jobs.jooq.generated.enums.JobStreamStatusRunState
import io.airbyte.server.repositories.StreamStatusesRepository
import io.airbyte.server.repositories.StreamStatusesRepository.FilterParams
import io.airbyte.server.repositories.domain.StreamStatus
import io.airbyte.server.repositories.domain.StreamStatus.StreamStatusBuilder
import io.airbyte.server.repositories.domain.StreamStatusRateLimitedMetadataRepositoryStructure
import jakarta.inject.Singleton
import java.time.Instant
import java.time.OffsetDateTime
import java.time.ZoneOffset
import java.util.Locale

/**
 * Maps between the API and Persistence layers. It is not static to be injectable and enable easier
 * testing in dependents.
 */
@Singleton
class StreamStatusesMapper {
  // API to Domain
  fun map(api: StreamStatusCreateRequestBody): StreamStatus {
    val domain =
      StreamStatusBuilder()
        .runState(map(api.runState))
        .transitionedAt(fromMills(api.transitionedAt))
        .workspaceId(api.workspaceId)
        .connectionId(api.connectionId)
        .jobId(api.jobId)
        .jobType(map(api.jobType))
        .attemptNumber(api.attemptNumber)
        .streamNamespace(api.streamNamespace)
        .streamName(api.streamName)

    if (null != api.incompleteRunCause) {
      domain.incompleteRunCause(map(api.incompleteRunCause))
    }

    if (null != api.metadata) {
      domain.metadata(map(api.metadata))
    }

    return domain.build()
  }

  fun map(api: StreamStatusUpdateRequestBody): StreamStatus {
    val domain =
      StreamStatusBuilder()
        .runState(map(api.runState))
        .transitionedAt(fromMills(api.transitionedAt))
        .workspaceId(api.workspaceId)
        .connectionId(api.connectionId)
        .jobId(api.jobId)
        .jobType(map(api.jobType))
        .attemptNumber(api.attemptNumber)
        .streamNamespace(api.streamNamespace)
        .streamName(api.streamName)
        .id(api.id)

    if (null != api.incompleteRunCause) {
      domain.incompleteRunCause(map(api.incompleteRunCause))
    }

    if (null != api.metadata) {
      domain.metadata(map(api.metadata))
    }

    return domain.build()
  }

  fun map(apiEnum: StreamStatusJobType?): JobStreamStatusJobType? =
    if (apiEnum != null) JobStreamStatusJobType.lookupLiteral(apiEnum.name.lowercase(Locale.getDefault())) else null

  fun map(apiEnum: StreamStatusRunState?): JobStreamStatusRunState? =
    if (apiEnum != null) JobStreamStatusRunState.lookupLiteral(apiEnum.name.lowercase(Locale.getDefault())) else null

  fun map(apiEnum: StreamStatusIncompleteRunCause?): JobStreamStatusIncompleteRunCause? =
    if (apiEnum != null) JobStreamStatusIncompleteRunCause.lookupLiteral(apiEnum.name.lowercase(Locale.getDefault())) else null

  fun map(api: Pagination?): StreamStatusesRepository.Pagination? =
    if (api != null) {
      StreamStatusesRepository.Pagination(
        api.rowOffset / api.pageSize,
        api.pageSize,
      )
    } else {
      null
    }

  fun map(api: StreamStatusListRequestBody): FilterParams =
    FilterParams(
      api.workspaceId,
      api.connectionId,
      api.jobId,
      api.streamNamespace,
      api.streamName,
      api.attemptNumber,
      map(api.jobType),
      map(api.pagination),
    )

  // Domain to API
  fun map(domain: StreamStatus): StreamStatusRead {
    val api = StreamStatusRead()
    api.id = domain.id
    api.connectionId = domain.connectionId
    api.workspaceId = domain.workspaceId
    api.jobId = domain.jobId
    api.jobType = map(domain.jobType)
    api.attemptNumber = domain.attemptNumber
    api.streamName = domain.streamName
    api.streamNamespace = domain.streamNamespace
    api.transitionedAt = domain.transitionedAt!!.toInstant().toEpochMilli()
    api.runState = map(domain.runState)

    if (null != domain.incompleteRunCause) {
      api.incompleteRunCause = map(domain.incompleteRunCause)
    }

    if (null != domain.metadata) {
      api.metadata = map(domain.metadata)
    }

    return api
  }

  fun map(domainEnum: JobStreamStatusJobType?): StreamStatusJobType? =
    if (domainEnum != null) StreamStatusJobType.fromValue(domainEnum.name.uppercase(Locale.getDefault())) else null

  fun map(domainEnum: JobStreamStatusRunState?): StreamStatusRunState? =
    if (domainEnum != null) StreamStatusRunState.fromValue(domainEnum.name.uppercase(Locale.getDefault())) else null

  fun map(domainEnum: JobStreamStatusIncompleteRunCause?): StreamStatusIncompleteRunCause? =
    if (domainEnum != null) StreamStatusIncompleteRunCause.fromValue(domainEnum.name.uppercase(Locale.getDefault())) else null

  fun map(rateLimitedMetadata: JsonNode): StreamStatusRateLimitedMetadata {
    val rateLimitedInfo =
      Jsons.`object`(
        rateLimitedMetadata,
        StreamStatusRateLimitedMetadataRepositoryStructure::class.java,
      )
    return StreamStatusRateLimitedMetadata().quotaReset(rateLimitedInfo.quotaReset)
  }

  fun map(rateLimitedMetadata: StreamStatusRateLimitedMetadata): JsonNode {
    val streamStatusRateLimitedMetadataRepositoryStructure =
      StreamStatusRateLimitedMetadataRepositoryStructure(rateLimitedMetadata.quotaReset)
    return Jsons.jsonNode(streamStatusRateLimitedMetadataRepositoryStructure)
  }

  fun fromMills(millis: Long): OffsetDateTime = OffsetDateTime.ofInstant(Instant.ofEpochMilli(millis), ZoneOffset.UTC)
}
